packages/babel-helpers/src/helpers.js

```diff
@@ -633,28 +603,91 @@ helpers.possibleConstructorReturn = () => template.program.ast`
   }
 `;
 
+helpers.superPropBase = () => template.program.ast`
+  import getPrototypeOf from "getPrototypeOf";
+
+  export default function _superPropBase(object, property) {
+    // Yes, this throws if object is null to being with, that's on purpose.
+    while (!Object.prototype.hasOwnProperty.call(object, property)) {
+      object = getPrototypeOf(object);
+      if (object === null) break;
+    }
+    return object;
+  }
+`;
+
+helpers.get = () => template.program.ast`
+  import getPrototypeOf from "getPrototypeOf";
+  import superPropBase from "superPropBase";
+
+  export default function _get(object, property, receiver) {
+    var base = superPropBase(object, property);
+
+    if (!base) return;
+
+    var desc = Object.getOwnPropertyDescriptor(base, property);
+    if (desc.get) {
+      return desc.get.call(receiver);
+    } else if ("value" in desc) {
+      return desc.value;
+    }
+  }
+`;
+
+helpers.isStrict = () => template.program.ast`
+  export default function _isStrict() {
+    var strict = false;
+    try {
+      var obj = { get test() {} };
+      obj.test = 1;
+    } catch (e) {
+      strict = true;
+    }
+    _isStrict = function() { return strict; };
+    return strict;
+  }
+`;
+
 helpers.set = () => template.program.ast`
   import getPrototypeOf from "getPrototypeOf";
+  import superPropBase from "superPropBase";
+  import isStrict from "isStrict";
+  import defineProperty from "defineProperty";
 
   export default function _set(object, property, value, receiver) {
-    var desc = Object.getOwnPropertyDescriptor(object, property);
-
-    if (desc === undefined) {
-      var parent = getPrototypeOf(object);
+    var base = superPropBase(object, property);
+    var desc;
 
-      if (parent !== null) {
-        _set(parent, property, value, receiver);
+    if (base) {
+      desc = Object.getOwnPropertyDescriptor(base, property);
+      if (desc.set) {
+        desc.set.call(receiver, value);
+        return value;
+      } else if (desc.get || !desc.writable) {
+        // this will throw an error in strict code, and will silently fail in
+        // non-strict.
+        return base[property] = value;
       }
-    } else if ("value" in desc && desc.writable) {
-      desc.value = value;
-    } else {
-      var setter = desc.set;
+    }
 
-      if (setter !== undefined) {
-        setter.call(receiver, value);
+    // Without a super that defines the property, spec boils down to "define on
+    // receiver" for some reason.
+    desc = Object.getOwnPropertyDescriptor(receiver, property);
+    if (desc) {
+      if (desc.set) {
+        if (isStrict()) {
```

Comment:
Hmm, I wonder if it'd be better to actually pass this as a true/false value to the helper? Since it is the caller's location, not the actual helper that we want to know the strictness of. Especially with https://github.com/babel/babel/pull/7411 and if we do `sourceType:script` I'd expect a lot more code to be strict only inside class bodies for instance, so this helper would return the wrong value.

Reply:
That'd be great. How do I get that value? The same traversal?

Reply:
Done.